### PR TITLE
Fixes for nominal geometry.

### PR DESCRIPTION
**Pull request by sophiemiddleton**
This is the first round of upgrades to how we extract the relative positions of the DS/CRV components.

- Will do the extracted version in a subsequent PR.
- IPA/OPA for some reason cannot be done yet (I dont understand their heirarchy)

### CODE REVIEW

File: src/REveMu2eMainWindow.cc [hunk 1]
```diff
@@ -252,84 +149,233 @@ void REveMu2eMainWindow::GeomDrawer(TGeoNode* node, REX::REveTrans& trans, REX::
...
+      ctrans.SetFrom(trans.Array());
+      {
+        TGeoMatrix     *gm = n->GetMatrix();
+        const Double_t *rm = gm->GetRotationMatrix();
+        const Double_t *tv = gm->GetTranslation();
+        REX::REveTrans t;
+        t(1,1) = rm[0]; t(1,2) = rm[1]; t(1,3) = rm[2];
+        t(2,1) = rm[3]; t(2,2) = rm[4]; t(2,3) = rm[5];
+        t(3,1) = rm[6]; t(3,2) = rm[7]; t(3,3) = rm[8];
+        t(1,4) = tv[0]  ; t(2,4) = tv[1]  ; t(3,4) = tv[2] ;
```
Comment (brownd1978) [inline src/REveMu2eMainWindow.cc:198]: Add a brief comment what these lines do.

File: src/REveMu2eMainWindow.cc [hunk 2]
```diff
@@ -252,84 +149,233 @@ void REveMu2eMainWindow::GeomDrawer(TGeoNode* node, REX::REveTrans& trans, REX::
...
-  }
-
-  if(geomOpt.showSTM){
-    static std::vector <std::string> substrings_stm  {"stmDet1Can","stmDet1","stmDet2Can","stmDet2"};
-    for(auto& i: substrings_stm){
-      shift.at(0) = 0;
-      shift.at(1) = 0;
-      shift.at(2) = 0;
-      showNodesByName(node,i,kFALSE, 0, trans, beamlineholder, maxlevel, level, false, false, shift, false, false, drawconfigf.getInt("BLColor"));
+    /*if(geomOpt.showST and !geomOpt.extracted){
```
Comment (brownd1978) [inline src/REveMu2eMainWindow.cc:360]: What is the plan for this commented code?

**Review by brownd1978 — COMMENTED** [review]
In general this looks like a big improvement. Can you verify that the visual output still puts the components in the same place as before?
There are still a few hard-coded numbers, can you comment if they are still needed? Are they related to the email comment about IPA and OPA placement?

File: inc/REveMu2eMainWindow.hh [hunk 3]
```diff
@@ -147,8 +149,12 @@ namespace mu2e {
...
+            void getOffsets(TGeoNode* n, const std::string& str, REX::REveTrans& trans, int maxlevel, int level, std::vector<std::pair<std::string, std::vector<float>>> & offsets);
             void changeNodesByName(TGeoNode* n, const std::string& str, bool onOff, int _diagLevel, REX::REveTrans& trans,  REX::REveElement* holder, int maxlevel, int level, bool caloshift, bool crystal, std::vector<double> shift, bool print, bool single);
             void GeomDrawer(TGeoNode* node, REX::REveTrans& trans,  REX::REveElement* bholder, REX::REveElement* trholder, REX::REveElement* cholder,REX::REveElement* crholder, REX::REveElement* vholder, REX::REveElement* tholder,int maxlevel, int level, GeomOptions geomOpts);
+
+            void GeomDrawerDS(TGeoNode* node, REX::REveTrans& trans, REX::REveElement* bholder, REX::REveElement* trholder, REX::REveElement* cholder,REX::REveElement* crholder, REX::REveElement* vholder, REX::REveElement* tholder, int maxlevel, int level, GeomOptions geomOpt, std::vector<std::pair<std::string, std::vector<float>>>& offsets);
```
[the remark this answers is not in the document]
Comment (sophiemiddleton) [reply inc/REveMu2eMainWindow.hh:156]: they are modified, things are added to them, they hold our geom scenes

File: inc/REveMu2eMainWindow.hh [hunk 4]
```diff
@@ -147,8 +149,12 @@ namespace mu2e {
...
+            void GeomDrawerSol(TGeoNode* node, REX::REveTrans& trans, REX::REveElement* beamlineholder, int maxlevel, int level, GeomOptions geomOpt, std::vector<std::pair<std::string, std::vector<float>>>& offsets);
```
[the remark this answers is not in the document]
Comment (sophiemiddleton) [reply inc/REveMu2eMainWindow.hh:157]: Sol is all solenoids, DS is just the insides of the DS. DS could be called nominal (also has CRV)

File: src/REveMu2eMainWindow.cc [hunk 5]
```diff
@@ -5,39 +5,19 @@ namespace REX = ROOT::Experimental;
 using namespace std;
 using namespace mu2e;
 
-// Get geom information from configs:
-std::string calfilename("REve/config/geomutils.txt");
+//positions of components for DS component offsets
```
[the remark this answers is not in the document]
Comment (sophiemiddleton) [reply src/REveMu2eMainWindow.cc:8]: We dont need this anymore, its old

File: src/REveMu2eMainWindow.cc [hunk 6]
```diff
@@ -252,84 +149,233 @@ void REveMu2eMainWindow::GeomDrawer(TGeoNode* node, REX::REveTrans& trans, REX::
     static std::vector <std::string> substrings_ex {"CRSscintLayer_0","CRSmotherLayer_CRV_EX"};//
     shift.at(0) = 0;
     shift.at(1) = firstCounterEX[1]/10;
-    shift.at(2) =  82.8/2 + 124.2 + firstCounterEX[2]/10 - trackerz0_extracted/10; //tracker - first counter pos plus GDML position of first + wisth of module
-    //std::cout<<"Position of CRV "<<firstCounterEX[2] - trackerz0_extracted<<std::endl;
+    shift.at(2) =  82.8/2 + 124.2 + firstCounterEX[2]/10 - trackerz0_extracted/10;
```
[the remark this answers is not in the document]
Comment (sophiemiddleton) [reply src/REveMu2eMainWindow.cc:152]: yes....its the extracted...I havent got to it yet

File: src/REveMu2eMainWindow.cc (hunk 2 above)
Comment (sophiemiddleton) [reply src/REveMu2eMainWindow.cc:360]: As per the note, I'm not sure how to find the IPA/OPA from the GDML, its nested somewhere weirdly...I want to figure it out, but it comes next on my list after the CRV

**Comment by sophiemiddleton** [comment]
thanks, as for visual. Things mostly look the same. I'm still working to understand the exact placement in z of the CRV

**Comment by sophiemiddleton** [comment]
visuals look good, the missing pieces of IPA/OPA and extracted CRV positions remain to be finished

**Comment by sophiemiddleton** [comment]
If you are OK with this being an intermediate step I can merge it. I will try to get to the extracted part later this week or next. If you want to wait until then, I can leave this open. Let me know what you prefer

**Comment by brownd1978** [comment]
I’d like to be sure the IPA and ST are still in the correct positions. Can
you verify that? Then I’ll merge.

David Brown            ***@***.***
Office Phone (510) 486-7261
Lawrence Berkeley National Lab
M/S 50R5008 (50-6026C)  Berkeley, CA 94720

**Comment by sophiemiddleton** [comment]
The ST is in the same location as before.

The IPA has not being upgraded yet

**Comment by sophiemiddleton** [comment]
As mentioned, the proton abs are now correct, can I merge this?

**Comment by kutschke** [comment]
> As mentioned, the proton abs are now correct, can I merge this?

I am starting to review it now.

File: src/REveMu2eMainWindow.cc [hunk 7]
```diff
@@ -252,84 +150,259 @@ void REveMu2eMainWindow::GeomDrawer(TGeoNode* node, REX::REveTrans& trans, REX::
...
+      {
+        x_hall = x_world + offsets[i].second[0];
+        y_hall = y_world + offsets[i].second[1];
+        z_hall = z_world + offsets[i].second[2];
+      }
+      if(offsets[i].first.find("DS3Vacuum") != string::npos)      {
+        x_ds3 = x_hall + offsets[i].second[0];
+        y_ds3 = y_hall + offsets[i].second[1];
+        z_ds3 = z_hall + offsets[i].second[2];
+        z_crv = -1*offsets[i].second[2];//595.3
```
Comment (kutschke) [inline src/REveMu2eMainWindow.cc:304]: These comments  may not age well.  If there is a good reason to keep them, name a git tag or SHA for which they are correct.

File: src/REveMu2eMainWindow.cc [hunk 8]
```diff
@@ -252,84 +150,259 @@ void REveMu2eMainWindow::GeomDrawer(TGeoNode* node, REX::REveTrans& trans, REX::
...
+    getOffsets( pn, str, trans,  maxlevel, level, offsets);
+  }
+}
+
+void REveMu2eMainWindow::GeomDrawerSol(TGeoNode* node, REX::REveTrans& trans, REX::REveElement* beamlineholder, int maxlevel, int level, GeomOptions geomOpt, std::vector<std::pair<std::string, std::vector<float>>>& offsets) {
+
+  double x_sol = 0; double y_sol = 0; double z_sol = 0; 
+  double x_pt = 0; double y_pt = 0; double z_pt =0;
+  for(unsigned int i = 0; i < offsets.size(); i++){
+      if(offsets[i].first.find("DS3Vacuum0") != string::npos)      {
```
Comment (kutschke) [inline src/REveMu2eMainWindow.cc:242]: The "0" is not part of the volume name.  It is part of the uniqueness decoration added by G4's GDML writer.  Please remove it.

**Review by kutschke — CHANGES_REQUESTED** [review]
I am requesting 2 minor changes that should be quick.  See the comments about the "current crashing" comment and the superfluous "0" on the name DS3Vacuum.  The other comments are requests for improvements in a future PR.

File: config/drawutils.txt [hunk 9]
```diff
@@ -21,4 +21,4 @@ int CRVBarColor = 632;
 int RecoTrackColor = 600; //kGreen
 int CRVHitColor = 600; //kBlue
 int TrkHitColor = 632; //kBlue
-int  TrkNoHitColor = 600; //kBlue
+int TrkNoHitColor = 600; //kBlue
```
[the remark this answers is not in the document]
Comment (sophiemiddleton) [reply config/drawutils.txt:24]: Yes, good point, I already removed the other config

**Comment by sophiemiddleton** [comment]
thanks @kutschke I have implemented your suggestions

**Review by kutschke — APPROVED** [review]
Thanks for the fixes.  Please keep the other requests on your radar.

When the time comes, I can show you how to use symbolic names for the colors even in a simple config file.

**Review by brownd1978 — APPROVED** [review]
This looks good.

**Pull request merged**